// prisma-laravel-migrate/dist/generator/modeler/relationship/types.js
/* ------------------------------- Utils ----------------------------------- */
export const getModel = (dmmf, name) => dmmf.datamodel.models.find((m) => m.name === name);
export const dbNameOf = (m) => m.dbName ?? m.name;
export const conventionalPivotName = (a, b) => [a, b].map((s) => s.toLowerCase()).sort().join("_");
export const objRels = (m) => m.fields.filter((f) => f.kind === "object");
export const scalarNames = (m) => m.fields.filter((f) => f.kind === "scalar").map((s) => s.name);
export const PIVOT_SCALAR_WHITELIST = new Set([
    "id",
    "created_at",
    "updated_at",
    "deleted_at",
    "createdAt",
    "updatedAt",
    "deletedAt",
]);
export const getPrimaryKeyFields = (m) => {
    const pk = m.primaryKey?.fields;
    if (pk?.length)
        return pk;
    const ids = m.fields.filter((f) => f.isId).map((f) => f.name);
    return ids.length ? ids : [];
};
export const hasIntersection = (a = [], b = []) => {
    const S = new Set(a);
    return b.some((x) => S.has(x));
};
export const setsEqual = (a, b) => {
    if (a.length !== b.length)
        return false;
    const S = new Set(a);
    return b.every((x) => S.has(x));
};
export const getUniqueSets = (m) => {
    const sets = [];
    const uniqueFields = m.uniqueFields;
    if (uniqueFields?.length)
        sets.push(...uniqueFields);
    const uniqueIndexes = m.uniqueIndexes;
    if (uniqueIndexes?.length)
        sets.push(...uniqueIndexes.map((u) => u.fields));
    const pk = getPrimaryKeyFields(m);
    if (pk.length)
        sets.push(pk);
    return sets;
};
export const isUniqueOn = (m, fields) => {
    if (!fields.length)
        return false;
    const uniques = getUniqueSets(m);
    const sorted = [...fields].sort();
    return uniques.some((set) => setsEqual(set.slice().sort(), sorted));
};
