// prisma-laravel-migrate/dist/generator/migrator/rule-definition.js
import { rules as ruleList, defaultBuild } from "./rules.js";
/**
 * Encapsulates all special‐case column rendering rules.
 * Needs access to the full DMMF document and the set of all column definitions
 * in order to support rules that depend on other columns (e.g. composite keys).
 */
export class RuleResolver {
    dmmf;
    customRules;
    defaultMaps;
    definitions = [];
    constructor(dmmf, customRules = [], defaultMaps = {}) {
        this.dmmf = dmmf;
        this.customRules = customRules;
        this.defaultMaps = defaultMaps;
    }
    /**
     * Supply the full list of column definitions so that rules can inspect
     * other columns (e.g. detect composite primary keys or multi‐column morphs).
     */
    setDefinitions(defs) {
        this.definitions = defs;
    }
    rules = ruleList;
    /**
     * Returns the special‐case lines for this column,
     * or an empty array if none of the rules match.
     */
    resolve(def) {
        for (const rule of [...this.rules, ...this.customRules]) {
            if (rule.utility)
                continue; // skip utility rules here
            if (rule.test(def, this.definitions, this.dmmf)) {
                return rule.render(def, this.definitions, this.dmmf, this.defaultMaps);
            }
        }
        return defaultBuild(def, this.defaultMaps); // fallback
    }
    /** NEW: run all *utility* rules once and gather their snippets */
    resolveUtilities() {
        const snippets = [];
        for (const rule of [...this.rules, ...this.customRules]) {
            if (!rule.utility)
                continue;
            // many utility rules fire only once per model; iterate all defs
            for (const def of this.definitions) {
                if (rule.test(def, this.definitions, this.dmmf)) {
                    const { snippet } = rule.render(def, this.definitions, this.dmmf, this.defaultMaps);
                    snippets.push(...snippet);
                }
            }
        }
        return snippets;
    }
}
